emcc.py

```diff
@@ -22,40 +22,30 @@
 
 from tools.toolchain_profiler import ToolchainProfiler
 
-import base64
-import glob
-import hashlib
 import json
 import logging
 import os
 import re
 import shlex
-import shutil
-import stat
 import sys
 import time
 import tarfile
-from enum import Enum, unique, auto
+from enum import Enum, auto, unique
 from subprocess import PIPE
-from urllib.parse import quote
 
 
-import emscripten
-from tools import shared, system_libs, utils, ports, filelock
+from tools import shared, system_libs, utils, ports
 from tools import colored_logger, diagnostics, building
-from tools.shared import unsuffixed, unsuffixed_basename, WINDOWS, safe_copy
-from tools.shared import run_process, read_and_preprocess, exit_with_error, DEBUG
-from tools.shared import do_replace
+from tools.shared import unsuffixed, unsuffixed_basename, get_file_suffix
+from tools.shared import run_process, exit_with_error, DEBUG
+from tools.shared import in_temp, OFormat
+from tools.shared import DYNAMICLIB_ENDINGS, STATICLIB_ENDINGS
 from tools.response_file import substitute_response_files
-from tools.minimal_runtime_shell import generate_minimal_runtime_html
-import tools.line_endings
-from tools import feature_matrix
-from tools import js_manipulation
-from tools import webassembly
 from tools import config
 from tools import cache
+from tools import link
```

Comment:
IIUC, Python's imports are lazy like Node.js `require` and evaluated only when the `import` line is actually reached.

If so, I wonder if there's any perceptible performance difference if you move this import down to where it's actually needed, after we figured out that we're not just compiling. If there is, it could reduce the startup component of the Python cost breakdown you mentioned in the other issue.

Reply:
Yes, its worth trying.  But I'm leaving that kind of things for a followup.  This PR not focused on performance, just a pure refactor.